bbl_screen-patch/patches/printerui/qml/x1plus/GcodeGenerator.js

```diff
@@ -230,61 +145,73 @@ function M975(enabled = true) {
 
 /* k value */
 function M900(k, l, m) {
-    return this.createGcode('M900', {K: k, L: l, M: m});
+    return createGcode('M900', {K: k, L: l, M: m});
+}
+
+
+/**
+ * Interpolation of Bambu print speed parameters. Details about this process at here:
+ * https://github.com/jphannifan/x1plus-testing/blob/main/BL-speed-adjust.md
+ * We can modify this readme and turn it into Wiki content.
+ * */
+var speed_interp = {
+    speed_fraction: (speedPercentage) => { return Math.floor(10000/speedPercentage)/100},
+    acceleration_magnitude: (speedFraction) => {return Math.exp((speedFraction - 1.0191) / -0.814)},
+    feed_rate: (speedPercentage) => {return (0.00006426)*speedPercentage ** 2 + (-0.002484)*speedPercentage + 0.654},
+    level: (accelerationMagnitude) => {return (1.549 * accelerationMagnitude ** 2 - 0.7032 * accelerationMagnitude + 4.0834)}
+}
+
+/**
+ * Generates G-code based on the print speed level.
+ * 
+ * @param {number} speedPercentage The desired print speed as a percentage of the normal speed (100%).
+ *                                 Accepts: 50 (Silent), 100 (Normal), 125 (Sport), 166 (Luda)
+ * @returns {string} The G-code string to set print speed
+ */
+function speed(speedPercentage) {
+    if (speedPercentage <30 || speedPercentage > 180){
+        speedPercentage = 100;
+    }
+    // Convert percentage to a fraction, use Math.floor() to keep our target % the same as Bambu's reported %
+    var speedFraction = speed_interp.speed_fraction(speedPercentage);
+    
+    // Calculate acceleration magnitude from speed fraction based on log trendline
+    var accelerationMagnitude = speed_interp.acceleration_magnitude(speedFraction);
+    
+    // Interpolate feed rate from R
+    var feedRate = speed_interp.feed_rate(speedPercentage);
```

Comment:
I feel uneasy about this being how things are derived, if this is going to drive the GUI for how speed is calculated.  In this case, '100% speed' *will set the speed to 105%*!  Obviously we must not do that: that will break prints that are carefully calibrated to be at the VFR limit (which seems to be a thing users do...).

Reply:
I see your point, and it's a valid concern. 

But this is not something I would be putting up a PR for if I had any question about. The printer only reports even numbered percentages, but I've tested every single % from 30 to 180% on about 40 hours of prints. Also Wolf_with_sword added the same trendlines and a speed slider to his HA workflow so this is already extremely well tested. 

I have a writeup of the interpolation and a comparison of Bambu's parameters with curve fit data, which I can add to the wiki in a format that makes sense. I don't know what the answer to this question should be though. We could fix the values for the 4 different speed levels, but there isn't a way to fit all of the points easily. I'm willing to modify it if you think that's best, just need some suggestions.

Reply:
I do agree something needs to change, I'm just not sure what yet. I'll think on it